### PR TITLE
Integrate #3 of team #2

### PR DESCRIPTION
**Pull request by TobiasSchalau**

### CODE REVIEW

**Comment by TobiasSchalau** [comment]
I'm a little bit confused why black test is failing, because when I run ```black --check``` everything is fine. I already checked for the right version of black. 
Test failes in code of team #1 in "tests/attacks/class_test.py" and "tests/metrics/test_accuracy.py".

**Comment by travela** [comment]
Black works now. Regarding the tests: we see that you deleted them in your last sprint, but our last pull request was based on an older version and for some reason they were added again with our last PR. I'm assuming we can just delete them again?

**Comment by Erdnaf** [comment]
@TobiasSchalau Hey, actualy both of the test that are failing don't exist in the current version of our branch anymore. They were at some time not needed anymore. I think you can savely delete them.
The class test was only for the interface, to see if everything works and the other test was integrated into `basics_test.py` witch is in `team1sprint3` and will be merged when our PR #142 is merged

**Comment by TobiasSchalau** [comment]
@Erdnaf thank you for the response. I deleted the tests, but there are still 2 tests with errors. Below yu can see a snip from pytest.
grafik

**Comment by Erdnaf** [comment]
> @Erdnaf thank you for the response. I deleted the tests, but there are still 2 tests with errors. Below yu can see a snip from pytest.
> grafik

Hey, there i actualy dont know why taht is faling, maybe @jtorhoff from our tem can help?

**Comment by jtorhoff** [comment]
> > @Erdnaf thank you for the response. I deleted the tests, but there are still 2 tests with errors. Below yu can see a snip from pytest.
> > grafik
> 
> Hey, there i actualy dont know why that is faling, maybe @jtorhoff from our team can help?

Looking through the logs shows a segfault: `RuntimeError: DataLoader worker (pid 2877) is killed by signal: Segmentation fault.`. My guess is that the tiny VM that runs the tests runs out of memory. Not sure why it doesn't happen with other pull requests, though. In any case I think we have to make our tests run on less data. Right now the tests need >15 min and sometimes up to half an hour.

```
The above exception was the direct cause of the following exception:

    def test_membership_inference_attack():
>       x_train, y_train, x_test, y_test = CIFAR10.numpy()

tests/attacks/membership_inference/membership_inference_test.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
privacy_evaluator/datasets/cifar10.py:93: in numpy
    x_train, y_train = next(iter(train_loader))
```

There might be some code that prevents garbage collection somewhere else, like a static class holding a reference to some object which takes a huge amount of memory. And once we reach this test it's impossible to allocate any more memory so the process is killed by the OS. That would be my guess.

@marisanest what's your take on this?

**Comment by jtorhoff** [comment]
Hey maybe you could try to simplify the `test_property_inference_attack`. You could, for example, only use a subset of the dataset, because I tried to run it on my laptop and it took 20 mins. I'm quite sure it has something to do with the memory usage, as `pytest` keeps a lot of stuff in memory until the summary report is created at the end. I also didn't encounter the issue when I ran all the tests on my laptop.

We will also try to simplify our tests to avoid unnecessary resources usage.

**Comment by marisanest** [comment]
Hey, thanks @jtorhoff for your detailed answer. I totally agree with your conclusion that the current memory usage is a problem. We should definitely investigate better ways to use the datasets (e.g. not initializing them over and over again or using the whole dataset for testing). I think we should also keep an eye on the size of the models we initialize for each test. These shouldn't be too big either and also should only be loaded once for all test. 

In our current sprint branch [team1sprint4](https://github.com/privML/privacy-evaluator/tree/team1sprint4) we introduced so-called fixtures. This could be a way to deal with it. @jtorhoff also open an issue for this testing problem here: #146. So please have a look @team2.

File: privacy_evaluator/models/torch/fc_neural_net.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import torch
+from torch import nn
+from typing import Tuple, Optional
+
+
+def FCNeuralNet(
```
Comment (fraboeni) [inline privacy_evaluator/models/torch/fc_neural_net.py:6]: Please also specify here that this is a model that can be used for testing the framework with data in the CIFAR or MNIST shape.

File: privacy_evaluator/models/tf/fc_neural_net.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from tensorflow import keras
+from tensorflow.keras.layers import Dense, Dropout, Flatten
+
+
+class FCNeuralNet(tf.Model):
+    """A simple fully-connected network for multi-classification.
+
+    Args:
+        num_classes: The number of classes involved in the classification.
+        dropout: Drop-out rate in the fully-connected layer.
```
[the remark this answers is not in the document]
Comment (yzchyx) [reply privacy_evaluator/models/tf/fc_neural_net.py]: Thank you, Franziska. We have changed that here as well as its torch-parallel.

File: privacy_evaluator/models/torch/cnn.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
+import torch
+from torch import nn
+from typing import Tuple, Union, Optional
+
+
+def ConvNet(
```
[the remark this answers is not in the document]
Comment (yzchyx) [reply privacy_evaluator/models/torch/cnn.py:6]: We have added "Note" in docstring to explain this is only compatible with the two shapes, both in `tf` and `torch` files

File: privacy_evaluator/models/torch/fc_neural_net.py (hunk 1 above)
Comment (yzchyx) [reply privacy_evaluator/models/torch/fc_neural_net.py:6]: In both `tf/fc_neural_net.py` and `torch/fc_neural_net.py` we have added "Note" in the second line of the docstring to explain it is just aimed at fetching a model for developers' test when a target model is required. Therefore, this is only compatible with the two shapes.

File: privacy_evaluator/utils/trainer.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+from tensorflow import keras
+import torch
+from torch import nn
+from torch.utils.data import DataLoader
+import numpy as np
+from typing import Tuple, Dict, Union
+from privacy_evaluator.utils.metric import cross_entropy_loss, accuracy
+
+
+def trainer(
```
[the remark this answers is not in the document]
Comment (yzchyx) [reply privacy_evaluator/utils/trainer.py:11]: Thank you for this instruction. Now we have added a quite detailed documentation for that.

[the remark this answers is not in the document]
Comment (fraboeni) [reply privacy_evaluator/utils/trainer.py:11]: @yzchyx the new version looks fantastic. Thank you for your efforts!

**Review by fraboeni — APPROVED** [review]
Thanks for revising my comments and the good implementation of them!

**Pull request merged**